src/js/relayclient/ServerHelper.js

```diff
@@ -0,0 +1,168 @@
+        this.pingedRelays = 0
+        this.relaysCount = filteredRelays.size
+    }
+
+    /**
+     * Ping those relays that were not returned yet. Remove the retuned relay (first to respond) from {@link remainingRelays}
+     * @returns the first relay to respond to a ping message. Note: will never return the same relay twice.
+     */
+    async nextRelay() {
+        if (this.filteredRelays.length === 0 || this.remainingRelays.length === 0) {
+            return null
+        }
+
+        let firstRelayToRespond = await this.raceToSuccess(
+            this.filteredRelays.map(relay => this.getRelayAddressPing(relay.relayUrl))
+        );
+        this.remainingRelays = this.remainingRelays.filter(a => a.relayUrl !== firstRelayToRespond.relayUrl)
+        this.pingedRelays++
+        return firstRelayToRespond
+    }
+
+    /**
+     * @returns JSON response from the relay server, but adds the requested URL to it:
+     * { relayUrl: url,
+     *   RelayServerAddress: address
+     * }
+     */
+    async getRelayAddressPing(relayUrl) {
+        let self = this
+        return new Promise(function (resolve, reject) {
+            let callback = function (error, body) {
+                if (error) {
+                    reject(error);
+                    return
+                }
+                try {
+                    body.relayUrl = relayUrl
+                    resolve(body);
+                }
+                catch (err) {
+                    reject(err);
+                }
+            }
+            self.httpSend.send(relayUrl + "/getaddr", {}, callback)
+        });
+    }
+
+    /**
+     * From https://stackoverflow.com/a/37235207 (modified to catch exceptions)
+     * Resolves once any promise resolves, ignores the rest, ignores rejections
+     */
+    async raceToSuccess(promises) {
+        let numRejected = 0;
+        return new Promise(
+            (resolve, reject) =>
+                promises.forEach(
+                    promise =>
+                        promise.then((res) =>
+                            resolve(res)
+                        )
+                            .catch(err => {
+                                // console.log(err)
+                                if (++numRejected === promises.length) {
+                                    reject("No response from any server.", err);
+                                }
+                            })
+                )
+        );
+    }
+}
+
+class ServerHelper {
+    /**
+     * 
+     * @param {*} minStake 
+     * @param {*} minDelay 
+     * @param {*} httpSend 
+     */
+    constructor(minStake, minDelay, httpSend) {
+        this.minStake = minStake
+        this.minDelay = minDelay
+        this.httpSend = httpSend
+
+        this.filteredRelays = []
+        this.isInitialized = false
+    }
+    /**
+     * @param {*} relayHub 
+     * @param {*} relayHubAddress 
+     */
+    setHub(relayHub) {
+        //todo: re-run only on hub
```

Comment:
the comment doesn't belong here, but to the `getPastEvents()`: its output should be cached until a hub is set (or: the cache should be cleared, when a new hub is set